btcec/signature.go

```diff
@@ -37,43 +36,32 @@ var (
 	oneInitializer = []byte{0x01}
 )
 
-// Serialize returns the ECDSA signature in the more strict DER format.  Note
-// that the serialized bytes returned do not include the appended hash type
-// used in Bitcoin signature scripts.
-//
-// encoding/asn1 is broken so we hand roll this output:
-//
-// 0x30 <length> 0x02 <length r> r 0x02 <length s> s
+// Serialize returns the ECDSA signature as r || s,
+// where both r and s are encoded into 32 byte big endian integers.
+// In order to remove malleability,
+// we set s = curve_order - s, if s is greater than curve.Order() / 2.
 func (sig *Signature) Serialize() []byte {
 	// low 'S' malleability breaker
 	sigS := sig.S
 	if sigS.Cmp(S256().halfOrder) == 1 {
 		sigS = new(big.Int).Sub(S256().N, sigS)
 	}
-	// Ensure the encoded bytes for the r and s values are canonical and
-	// thus suitable for DER encoding.
-	rb := canonicalizeInt(sig.R)
-	sb := canonicalizeInt(sigS)
-
-	// total length of returned signature is 1 byte for each magic and
-	// length (6 total), plus lengths of r and s
-	length := 6 + len(rb) + len(sb)
-	b := make([]byte, length)
-
-	b[0] = 0x30
-	b[1] = byte(length - 2)
-	b[2] = 0x02
-	b[3] = byte(len(rb))
-	offset := copy(b[4:], rb) + 4
-	b[offset] = 0x02
-	b[offset+1] = byte(len(sb))
-	copy(b[offset+2:], sb)
-	return b
+	rBytes := sig.R.Bytes()
+	sBytes := sigS.Bytes()
+	sigBytes := make([]byte, 64)
+	// 0 pad the byte arrays from the left if they aren't big enough.
+	copy(sigBytes[32-len(rBytes):32], rBytes)
+	copy(sigBytes[64-len(sBytes):64], sBytes)
+	return sigBytes
 }
 
 // Verify calls ecdsa.Verify to verify the signature of hash using the public
-// key.  It returns true if the signature is valid, false otherwise.
+// key. It returns true if the signature is valid, false otherwise.
+// This fails on any signature with signatures in non-canonical form.
 func (sig *Signature) Verify(hash []byte, pubKey *PublicKey) bool {
+	if sig.S.Cmp(S256().halfOrder) > 0 {
```

Comment:
It's confusing that on verification a strictly greater immediately implies an invalid signature while on decoding / parsing a greater or equal fails (see `parseSig`):
```go
if signature.S.Cmp(S256().halfOrder) >= 0 {
	return nil, errors.New("signature S is >= (curve.N) / 2")
}
```

Reply:
Thanks for pointing this out!